### PR TITLE
feat(relay): add some basic prometheus metrics

### PR DESCRIPTION
**Pull request by thomaseizinger**
The metrics are available at `http://{listen_addr}:8080/metrics`. Currently, we collect the following:

- Number of active allocations: We can have an alert once the number of allocations passes through a certain threshold.
- Outcome (success / error) and message kind (allocation / channel_bind / ...) of all responses: Summing all of these up would give you the total number of requests handled. We might want to have a grafana alert for an increased number of error responses.
- Total number of bytes relayed: Dividing this by time gives us an average "internal" bandwidth.

This is just a start, we can explore what else is useful as we have it operate.

Depends-On: https://github.com/firezone/firezone/pull/1743

### CODE REVIEW

**Comment by thomaseizinger** [comment]
> The prometheus endpoint can also operate as a basic health check. If that one is reachable, then the relay is still operating.

cc @AndrewDryga I think you asked for a health check endpoint somewhere.

**Comment by AndrewDryga** [comment]
@thomaseizinger would the metrics endpoint go down if the main logic doesn't work though? (Eg. if your auth token to the panel is invalid or there are some internal errors?)

**Comment by thomaseizinger** [comment]
> @thomaseizinger would the metrics endpoint go down if the main logic doesn't work though? (Eg. if your auth token to the panel is invalid or there are some internal errors?)

So far, any major state inconsistency or really bad error just exits the relay and thus the metrics endpoint with it.

To be frank, I think the better option is to use a supervising system like systemd.

What do you think of that strategy?

**Comment by AndrewDryga** [comment]
@thomaseizinger let's do it with a binary, I think we will want to deploy without a Docker for efficiency purposes anyways.

**Comment by thomaseizinger** [comment]
> @thomaseizinger let's do it with a binary, I think we will want to deploy without a Docker for efficiency purposes anyways.

Okay, I've removed the line about the health-check from the PR description.

**Comment by thomaseizinger** [comment]
@AndrewDryga There seems to be a compile error in the elixir code path of the integration test.

**Comment by AndrewDryga** [comment]
@thomaseizinger can you please rebase to `main`? I believe it should be fixed there.

File: rust/relay/src/main.rs [hunk 1]
```diff
@@ -68,7 +69,13 @@ async fn main() -> Result<()> {
         tracing_subscriber::fmt().with_env_filter(env_filter).init()
     }
 
-    let server = Server::new(args.public_ip4_addr, make_rng(args.rng_seed));
+    let mut metric_registry = Registry::with_prefix("relay");
+
+    let server = Server::new(
```
Comment (conectado) [inline rust/relay/src/main.rs:74]: We should have an argument to like `--serve-metrics` to enable/disable this.

**Review by conectado — APPROVED** [review]
I wonder if we could somehow use tracing for this using a crate like [this](https://crates.io/crates/tracing-opentelemetry) instead of introducing its own system?

Otherwise it LGTM just left a minor comment.

**Comment by jamilbk** [comment]
@thomaseizinger Looks like some PRs made it in before this one -- needs a rebase / conflict resolve.

**Comment by thomaseizinger** [comment]
> I wonder if we could somehow use tracing for this using a crate like [this](https://crates.io/crates/tracing-opentelemetry) instead of introducing its own system?

It doesn't quite achieve the same thing, does it? I've not worked with `tracing-opentelemetry` before and I am open to alternatives but from reading the docs it seems like it is not that easy to expose a few simple counters.

I think that there might be cleaner way of integrating these metrics into the relay server. Overall, I think that there is a lot of opportunity to refactor the relay into more distinct modules. But, I think it will take time and experience in operating it to see which abstractions are useful and where we need to invest to squash the bugs.

File: rust/relay/src/main.rs (hunk 1 above)
Comment (thomaseizinger) [reply rust/relay/src/main.rs:74]: I am curious as to why?

The metrics are collected any way (and the crate is designed for that to be extremely cheap) so disabling the server won't have a performance benefit. If you don't want to use the metrics, simply don't poll the endpoint.

If you don't want it exposed, don't forward the port.

I am a bit hesitant to adding configuration options for everything unless we can make a good case as to why.

Comment (thomaseizinger) [reply rust/relay/src/main.rs:74]: I am resolving this to merge this PR, we can always add a flag in a follow-up PR.

**Pull request merged**